src/Database/Statement.php

```diff
@@ -0,0 +1,370 @@
+        if ($row === false) {
+            return false;
+        }
+
+        if ($this->typeConverter !== null) {
+            return ($this->typeConverter)($row);
+        }
+
+        return $row;
+    }
+
+    /**
+     * Fetches the next row from a result set using PDO::FETCH_ASSOC
+     * and converts fields to types based on TypeMap.
+     *
+     * This behaves the same as `PDOStatement::fetch()` except an
+     * empty array is returned instead of false.
+     *
+     * @return array
+     */
+    public function fetchAssoc(): array
+    {
+        return $this->fetch('assoc') ?: [];
+    }
+
+    /**
+     * Fetches the next row from a result set using PDO::FETCH_num
+     * and converts fields to types based on TypeMap.
+     *
+     * This behaves the same as `PDOStatement::fetch()` except only
+     * a specific column from the row is returned.
+     *
+     * @param int $position Column index in result row.
+     * @return mixed
+     */
+    public function fetchColumn(int $position): mixed
+    {
+        $row = $this->fetch(PDO::FETCH_NUM);
+        if ($row && isset($row[$position])) {
+            return $row[$position];
+        }
+
+        return false;
+    }
+
+    /**
+     * Fetches the remaining rows from a result set
+     * and converts fields to types based on TypeMap.
+     *
+     * This behaves the same as `PDOStatement::fetchAll()`.
+     *
+     * @param string|int $mode PDO::FETCH_* constant or fetch mode name.
+     *   Valid names are 'assoc', 'num' or 'obj'.
+     * @return array
+     * @throws \InvalidArgumentException
+     */
+    public function fetchAll(string|int $mode = PDO::FETCH_NUM): array
+    {
+        $mode = $this->convertMode($mode);
+        $rows = $this->statement->fetchAll($mode);
+
+        if ($this->typeConverter !== null) {
+            return array_map($this->typeConverter, $rows);
+        }
+
+        return $rows;
+    }
+
+    /**
+     * Converts mode name to PDO constant.
+     *
+     * @param string|int $mode Mode name or PDO constant
+     * @return int
+     * @throws \InvalidArgumentException
+     */
+    protected function convertMode(string|int $mode): int
+    {
+        if (is_int($mode)) {
+            // We don't try to validate the PDO constants
+            return $mode;
+        }
+
+        static $MODES = ['assoc' => PDO::FETCH_ASSOC, 'num' => PDO::FETCH_NUM, 'obj' => PDO::FETCH_OBJ];
+        $mode = $MODES[$mode] ?? null;
+        if ($mode !== null) {
+            return $mode;
+        }
+
+        throw new InvalidArgumentException("Invalid fetch mode requested. Expected 'assoc', 'num' or 'obj'.");
+    }
+
+    /**
+     * Returns the number of rows affected by the last SQL statement.
+     *
+     * This behaves the same as `PDOStatement::rowCount()`.
+     *
+     * @return int
+     * @link https://www.php.net/manual/en/pdostatement.rowcount.php
+     */
+    public function rowCount(): int
+    {
+        if ($this->typeConverter !== null) {
```

Comment:
So this will effectively prevent use of `rowCount()` for SELECT queries even for dbs that support it isn't it? Not sure if we should do that. It could still be useful for debugging.

Reply:
I don't believe it's ever supported that way, but we can do that.

Reply:
Well `rowCount()` does work for all (most) dbs besides Sqlite even for SELECT and is shown in DebugKit currently.

Reply:
Fixed.